csrc/device_lower/validation.cpp

```diff
@@ -303,71 +302,89 @@ class ExprValidator : public OptOutDispatch {
-      auto producer_grouped_id_it = c2p_replay.getReplay().find(grouped_id);
-      NVF_ERROR(
-          producer_grouped_id_it != c2p_replay.getReplay().end(),
-          "No corresponding producer ID for the grouped consumer ID found: ",
-          grouped_id->toString());
-      auto producer_grouped_id = producer_grouped_id_it->second;
-
-      NVF_ERROR(
-          std::ranges::find(inp_tv->getLoopDomain(), producer_grouped_id) !=
-              inp_tv->getLoopDomain().end(),
-          "Corresponding grouped producer ID is not a loop ID: ",
-          producer_grouped_id->toString(),
-          " of ",
-          inp_tv->toString());
+    // Even if grouping is not used, CudaKernelGenerator assumes these
+    // properties for simplicity
 
-      NVF_ERROR(
-          ir_utils::mayRequireAllocation(inp_tv, producer_grouped_id),
-          "The corresponding producer loop ID for grouping must be actualy "
-          "allocated without parallelization");
+    // Both input and output must be Local tensors so that the op can be
+    // executed without explicit predication. This makes it simpler to
+    // generate code for grouped calls
+    NVF_ERROR_EQ(inp_tv->getMemoryType(), MemoryType::Local);
+    NVF_ERROR_EQ(out_tv->getMemoryType(), MemoryType::Local);
 
-      // Make sure all inner loop IDs should not contribute to the
-      // allocation
-      validateUnitStride(inp_tv, inp_tv->getLoopDomain(), producer_grouped_id);
+    // If not grouped, no more validation to do
+    if (grouped_id == nullptr) {
+      return;
     }
+
+    // The input will be initialized for this op. To avoid any
+    // potential conflict of initialization, require the input to be
+    // exclusively used by the grouped op.
+    NVF_ERROR_EQ(
+        inp_tv->uses().size(), 1, "Invalid tensor uses: ", inp_tv->toString());
+
+    // If it's grouped, it must correspond to the innermost subregion
+    // of the logical ID to group
+    IterDomain* logical_id_to_group =
+        out_tv->getLogicalDomain().at(logical_dim_to_group);
+    NVF_ERROR(
+        isInnermost(logical_id_to_group, grouped_id),
+        "Invalid ID to group: ",
+        grouped_id->toString(),
+        " of ",
+        out_tv->toString());
+
+    // The output is allocated on registers, so the allocation
+    // domain should be just the same as the loop domain. The
+    // grouped ID must have the unit stride, which means all of the
+    // inner IDs must not contribute to the allocation
+    validateUnitStride(out_tv, out_tv->getLoopDomain(), grouped_id);
+
+    // All of the inputs per thread must be provided to the device
+    // function as a contiguous chunk of memory where each element
+    // has a unit stride within its allocated buffer. More
+    // concretely, the input for argsort would be passed to the device
+    // function as follows:
+    //
+    //  // Each thread computes 8 argsort elements
+    //  float T1[8];
+    //  ...
+    //  blockArgsort(..., T1, ...);
```

Comment:
super nitpick: I think it's perfectly fine to use Argsort as an example to illustrate the idea. We can also just call it an abstract `GroupOp`

Reply:
Updated the comment